yoyodyne/models/modules/lstm.py

```diff
@@ -223,3 +223,149 @@ def forward(
+        """
+        # Encodes current symbol.
+        embedded = self.embed(symbol)
+        decoded, hiddens = self.module(embedded, last_hiddens)
+        # Gets emission probabilities over each hidden state (source symbol).
+        output = decoded.expand(-1, encoder_out.shape[1], -1)
+        output = torch.cat([output, encoder_out], dim=-1)
+        output = self.output_proj(output)
+        # Gets transition probabilities (alignment) for current states.
+        alignment = self._alignment_step(decoded, encoder_out, encoder_mask)
+        return base.ModuleOutput(
+            output=output, embeddings=alignment, hiddens=hiddens
+        )
+
+    def get_module(self) -> nn.LSTM:
+        return nn.LSTM(
+            self.embedding_size,
+            self.hidden_size,
+            num_layers=self.layers,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.bidirectional,
+        )
+
+    @property
+    def output_size(self) -> int:
+        return self.decoder_input_size + self.hidden_size
+
+    @property
+    def name(self) -> str:
+        return "hard attention LSTM"
+
+
+class ContextHardAttentionLSTMDecoder(HardAttentionLSTMDecoder):
+    # First order HMM variant of Hard Attention LSTM.
+    def __init__(self, *args, attention_context, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.delta = attention_context
+        self.alignment_proj = nn.Linear(
+            self.hidden_size * 2, (attention_context * 2) + 1
+        )  # Window size must include center and both sides.
+
+    def _alignment_step(
+        self,
+        decoded: torch.Tensor,
+        encoder_out: torch.Tensor,
+        encoder_mask: torch.Tensor,
+    ) -> torch.Tensor:
+        # Matrix multiply encoding and decoding for alignment representations.
+        # see: https://aclanthology.org/P19-1148/
+        # Expand decoded to concatenate with alignments
+        decoded = decoded.expand(-1, encoder_out.shape[1], -1)
+        # -> B x seq_len
+        alignment_scores = torch.cat(
+            [self.scale_encoded(encoder_out), decoded], dim=2
+        )
+        alignment_scores = self.alignment_proj(alignment_scores)
+        alignment_probs = nn.functional.softmax(alignment_scores, dim=-1)
+        # Limits context to window of self.delta (context length).
+        alignment_probs = alignment_probs.split(1, dim=1)
+        alignment_probs = torch.cat(
+            [
+                nn.functional.pad(
+                    t,
+                    (
+                        -self.delta + i,
+                        encoder_mask.shape[1] - (self.delta + 1) - i,
+                    ),
+                )
+                for i, t in enumerate(alignment_probs)
+            ],
+            dim=1,
+        )
+        # Get probability of alignments.
+        # Mask padding.
+        alignment_probs = alignment_probs * (~encoder_mask).unsqueeze(1) + 1e-7
```

Comment:
Can you make `1e-7` another module constant?

Reply:
Are we going to use it elsewhere?

Reply:
Maybe. (It is used in the model itself, sorta) But that's just one of the two concerns I have for whether something should be a constant. The other one is "is it inscrutable?" If so, putting it at the top and giving it a variable can help. A comment would also maybe help if you'd rather.

Reply:
It's moved to its own variable and added comment.  Let's keep here until we're seeing frequent use of epsilons in the library, then just move them all to default.